### Task 1: Generate and load the raw dataset
Load the CSV file data_safe_copy.csv into a DataFrame named tickets. Confirm that it has at least 500 rows. Print the first five rows and run info() to confirm the initial structure.

In [1]:
import pandas as pd
tickets=pd.read_csv("C:\\Users\\User\\Desktop\\Downloads\\data_safe_copy.csv")
tickets

,ticket_id,opened_at,category,priority,resolution_minutes
0,TK-0001,2024-03-01 09:00,billing,low,45
1,TK-0002,2024-03-01 09:30,account,medium,30
2,TK-0003,2024-03-01 10:00,technical,high,75
3,TK-0004,2024-03-01 10:30,billing,low,NaN
4,TK-0005,2024-03-01 11:00,account,medium,unknown
...,...,...,...,...,...
495,TK-0496,2024-03-01 11:30,TECHNICAL,high,60
496,TK-0497,2024-03-01 12:00,billing,high,15
497,TK-0498,2024-03-01 12:30,account,low,90
498,TK-0499,2024-03-01 13:00,technical,medium,120


500 rows × 5 columns

In [2]:
assert len(tickets)>=500,"Dataset has fewer than 500 rows!"
len(tickets)

500

In [3]:
print("\nFirst 5 rows:")
print(tickets.head())


First 5 rows:
  ticket_id         opened_at   category priority resolution_minutes
0   TK-0001  2024-03-01 09:00    billing      low                 45
1   TK-0002  2024-03-01 09:30    account   medium                 30
2   TK-0003  2024-03-01 10:00  technical     high                 75
3   TK-0004  2024-03-01 10:30    billing      low                NaN
4   TK-0005  2024-03-01 11:00    account   medium            unknown


In [4]:
tickets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   ticket_id           500 non-null    object
 1   opened_at           500 non-null    object
 2   category            500 non-null    object
 3   priority            500 non-null    object
 4   resolution_minutes  450 non-null    object
dtypes: object(5)
memory usage: 19.7+ KB


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   ticket_id           500 non-null    object
 1   opened_at           500 non-null    object
 2   category            500 non-null    object
 3   priority            500 non-null    object
 4   resolution_minutes  450 non-null    object
dtypes: object(5)
memory usage: 19.7+ KB

### Task 2: Parse dates and clean categories 
Convert opened_at to a datetime column. Then normalize category and priority by stripping whitespace and converting to lowercase. Show the unique values for both columns to confirm cleaning.

In [5]:
tickets["opened_at"]=pd.to_datetime(tickets["opened_at"],errors="coerce")
tickets

,ticket_id,opened_at,category,priority,resolution_minutes
0,TK-0001,2024-03-01 09:00:00,billing,low,45
1,TK-0002,2024-03-01 09:30:00,account,medium,30
2,TK-0003,2024-03-01 10:00:00,technical,high,75
3,TK-0004,2024-03-01 10:30:00,billing,low,NaN
4,TK-0005,2024-03-01 11:00:00,account,medium,unknown
...,...,...,...,...,...
495,TK-0496,2024-03-01 11:30:00,TECHNICAL,high,60
496,TK-0497,2024-03-01 12:00:00,billing,high,15
497,TK-0498,2024-03-01 12:30:00,account,low,90
498,TK-0499,2024-03-01 13:00:00,technical,medium,120


In [6]:
tickets["category"]=tickets["category"].str.strip().str.lower()
tickets["priority"] = tickets["priority"].str.strip().str.lower()
tickets

,ticket_id,opened_at,category,priority,resolution_minutes
0,TK-0001,2024-03-01 09:00:00,billing,low,45
1,TK-0002,2024-03-01 09:30:00,account,medium,30
2,TK-0003,2024-03-01 10:00:00,technical,high,75
3,TK-0004,2024-03-01 10:30:00,billing,low,NaN
4,TK-0005,2024-03-01 11:00:00,account,medium,unknown
...,...,...,...,...,...
495,TK-0496,2024-03-01 11:30:00,technical,high,60
496,TK-0497,2024-03-01 12:00:00,billing,high,15
497,TK-0498,2024-03-01 12:30:00,account,low,90
498,TK-0499,2024-03-01 13:00:00,technical,medium,120


In [7]:
tickets["category"].unique()

array(['billing', 'account', 'technical'], dtype=object)

array(['billing', 'account', 'technical'], dtype=object)


In [8]:
tickets["priority"].unique()

array(['low', 'medium', 'high'], dtype=object)

array(['low', 'medium', 'high'], dtype=object)


### Task 3: Convert resolution time safely
Convert resolution_minutes to numeric values. Invalid entries should become missing values. After conversion, compute the count of missing values and confirm it matches the number of intentionally introduced issues.

In [9]:
tickets["resolution_minutes"]=pd.to_numeric(tickets["resolution_minutes"],errors="coerce")
tickets

,ticket_id,opened_at,category,priority,resolution_minutes
0,TK-0001,2024-03-01 09:00:00,billing,low,45.0
1,TK-0002,2024-03-01 09:30:00,account,medium,30.0
2,TK-0003,2024-03-01 10:00:00,technical,high,75.0
3,TK-0004,2024-03-01 10:30:00,billing,low,NaN
4,TK-0005,2024-03-01 11:00:00,account,medium,NaN
...,...,...,...,...,...
495,TK-0496,2024-03-01 11:30:00,technical,high,60.0
496,TK-0497,2024-03-01 12:00:00,billing,high,15.0
497,TK-0498,2024-03-01 12:30:00,account,low,90.0
498,TK-0499,2024-03-01 13:00:00,technical,medium,120.0


In [10]:
missing_count=tickets["resolution_minutes"].isna().sum()
missing_count

np.int64(100)

np.int64(50)


### Task 4: Filter and validate cleaned data
Create a cleaned DataFrame named tickets_clean by dropping rows with missing resolution_minutes. Then verify:

All remaining resolution_minutes are non-negative
All categories and priorities are normalized
Show the number of records before and after cleaning.

In [11]:
total_records=tickets.shape[0]
print("Before cleaning number of tickets:",tickets.shape[0])

Before cleaning number of tickets: 500


Before cleaning number of tickets: 500


In [12]:
clean_tickets=tickets.dropna(subset=["resolution_minutes"])
print("After cleaning number of tickets:",clean_tickets.shape[0])

After cleaning number of tickets: 400


After cleaning number of tickets: 400


In [13]:
assert(clean_tickets["resolution_minutes"]>=0).all(),"All  negative resolution time found"
clean_tickets

,ticket_id,opened_at,category,priority,resolution_minutes
0,TK-0001,2024-03-01 09:00:00,billing,low,45.0
1,TK-0002,2024-03-01 09:30:00,account,medium,30.0
2,TK-0003,2024-03-01 10:00:00,technical,high,75.0
5,TK-0006,2024-03-01 11:30:00,technical,high,60.0
6,TK-0007,2024-03-01 12:00:00,billing,high,15.0
...,...,...,...,...,...
495,TK-0496,2024-03-01 11:30:00,technical,high,60.0
496,TK-0497,2024-03-01 12:00:00,billing,high,15.0
497,TK-0498,2024-03-01 12:30:00,account,low,90.0
498,TK-0499,2024-03-01 13:00:00,technical,medium,120.0


In [14]:
assert clean_tickets["category"].str.strip().str.lower().all()
print("All categories are normalized:",clean_tickets)

All categories are normalized:     ticket_id           opened_at   category priority  resolution_minutes
0     TK-0001 2024-03-01 09:00:00    billing      low                45.0
1     TK-0002 2024-03-01 09:30:00    account   medium                30.0
2     TK-0003 2024-03-01 10:00:00  technical     high                75.0
5     TK-0006 2024-03-01 11:30:00  technical     high                60.0
6     TK-0007 2024-03-01 12:00:00    billing     high                15.0
..        ...                 ...        ...      ...                 ...
495   TK-0496 2024-03-01 11:30:00  technical     high                60.0
496   TK-0497 2024-03-01 12:00:00    billing     high                15.0
497   TK-0498 2024-03-01 12:30:00    account      low                90.0
498   TK-0499 2024-03-01 13:00:00  technical   medium               120.0
499   TK-0500 2024-03-01 13:30:00    billing   medium                25.0

[400 rows x 5 columns]


All categories are normalized: [400 rows x 5 columns]


In [15]:
assert clean_tickets["priority"].str.strip().str.lower().all()
print("All priorities are normalized:",clean_tickets)

All priorities are normalized:     ticket_id           opened_at   category priority  resolution_minutes
0     TK-0001 2024-03-01 09:00:00    billing      low                45.0
1     TK-0002 2024-03-01 09:30:00    account   medium                30.0
2     TK-0003 2024-03-01 10:00:00  technical     high                75.0
5     TK-0006 2024-03-01 11:30:00  technical     high                60.0
6     TK-0007 2024-03-01 12:00:00    billing     high                15.0
..        ...                 ...        ...      ...                 ...
495   TK-0496 2024-03-01 11:30:00  technical     high                60.0
496   TK-0497 2024-03-01 12:00:00    billing     high                15.0
497   TK-0498 2024-03-01 12:30:00    account      low                90.0
498   TK-0499 2024-03-01 13:00:00  technical   medium               120.0
499   TK-0500 2024-03-01 13:30:00    billing   medium                25.0

[400 rows x 5 columns]


All priorities are normalized:[400 rows x 5 columns]


In [16]:
print("Validation successful.")

Validation successful.


### Task 5: Build a quality summary
Create a summary dictionary with keys total_records, clean_records, missing_resolution, avg_resolution, and max_resolution. Print the summary and verify that clean_records + missing_resolution equals total_records.

In [20]:
summary={
    " total_records":tickets.shape[0],
    "clean_records":clean_tickets.shape[0],
    "missing_resolution":tickets["resolution_minutes"].isna().sum(),
    "avg_resolution":clean_tickets["resolution_minutes"].mean(),
    "max_resolution":clean_tickets["resolution_minutes"].max(),
}
print(summary)

{' total_records': 500, 'clean_records': 400, 'missing_resolution': np.int64(100), 'avg_resolution': np.float64(57.5), 'max_resolution': 120.0}


{' total_records': 500, 'clean_records': 400, 'missing_resolution': np.int64(100), 'avg_resolution': np.float64(57.5), 'max_resolution': 120.0}

In [21]:
if summary["clean_records"] + summary["missing_resolution"] == summary["total_records"]:
    print("Verification successful: counts are consistent.")
else:
    print("Record count mismatch!")

KeyError: 'total_records'